rtla/cmd.sh

```diff
@@ -1,89 +1,264 @@
+        exit 1
+    fi
 done
 
-uname=`uname -nr`
-echo "$uname"
-rpm -q rtla
 
-cpulist=`get_allowed_cpuset`
-echo "allowed cpu list: ${cpulist}"
+uname=`uname -nr` # get the kernel version
+echo "############# dumping env ###########" | storage
+env | storage
+echo "#####################################" | storage
 
-# change list seperators from comma to new line and sort it
-cpulist=`convert_number_range ${cpulist} | tr , '\n' | sort -n | uniq`
+echo " " | storage
+echo "########## container info ###########" | storage
+echo "/proc/cmdline:" | storage
+cat /proc/cmdline | storage
+echo "#####################################" | storage
 
-declare -a cpus
-cpus=(${cpulist})
+echo "**** uid: $UID ****" | storage
 
-for cmd in rtla; do
-    command -v $cmd >/dev/null 2>&1 || { echo >&2 "$cmd required but not installed.  Aborting"; exit 1; }
-done
+release=$(cat /etc/os-release | sed -n -r 's/VERSION_ID="(.).*/\1/p')
 
-cyccore=${cpus[1]}
-cindex=2
-ccount=1
-while (( $cindex < ${#cpus[@]} )); do
-    cyccore="${cyccore},${cpus[$cindex]}"
-    cindex=$(($cindex + 1))
-    ccount=$(($ccount + 1))
-done
+# Check the mode
+case "$COMMAND" in
+  "timerlat"|"hwnoise"|"osnoise")
+    echo "Operating in $COMMAND mode." | storage
+    ;;
+  *)
+    echo "Error: Invalid mode. Please set COMMAND to 'timerlat', 'hwnoise', or 'osnoise'." | storage
+    exit 1
+    ;;
+esac
+
+cpuset=`get_allowed_cpuset`
+echo "cpuset: $cpuset" | storage
 
-sibling=`cat /sys/devices/system/cpu/cpu${cpus[0]}/topology/thread_siblings_list | awk -F '[-,]' '{print $2}'`
-if [[ "${sibling}" =~ ^[0-9]+$ ]]; then
-    echo "removing cpu${sibling} from the cpu list because it is a sibling of cpu${cpus[0]} which will be the cpu-main-thread"
-    cyccore=${cyccore//,$sibling/}
+## ALLCPUSET handles a bug in rhel9 (and possible others) where cpuset is 0 when no cpuset is defined.
+## Deprecate this once the bug is fixed.
+if [[ "$cpuset" == "0" && "$ALLCPUSET" == "y" ]]; then
+    echo "present cpus: " | storage
+    work_cores=`get_host_cores`
+    echo "Info: No cpu-set used, using all CPUs" | storage
+    echo "WARNING: If trying to run on CPU0, make sure ALLCPUSET=n (Default behavior)" | storage
+else
+    if [[ "$cpuset" -eq 0 ]]; then
+        echo "Warning: cpuset is bound to CPU0. If you are trying to use all cpus add '-e ALLCPUSET=y' to your run command." | storage
+    fi
+    if [[ "$ALLCPUSET" == "y" ]]; then
+        echo "Warning: ALLCPUSET is set to y but cpuset was provided. ALLCPUSET will be ignored in favor of the provided cpuset." | storage
+    fi
+    work_cores=$cpuset
 fi
-echo "new cpu list: ${cyccore}"
 
-if [ ${COMMAND} == "timerlat" ] || [ ${COMMAND} == "osnoise" ]; then
-    hist="hist"
+
+if [[ $HOUSEKEEPING_PER_NUMA -gt 0 ]]; then
+    populate_numa_info $cpuset
+    echo "Housekeeping core: $housekeeping_cores"
+    echo "Work cores: $work_cores"
+fi
```

Comment:
Why do we need more than one housekeeping core and why do they need to be on different numa nodes?

Reply:
We dont. Let me remove the second one.

Reply:
But then all the HOUSEKEEPING_PER_NUMA related changes you made are unnecessary - right?

Reply:
Ill try to explain my rationale here, but I do realize this may not be needed for your usecase. I believe the approach depends on your goals with the container. I've been discussing this with Clark Williams. We typically advise customers to adopt this approach during CPU partitioning certification. There are also situations where we don't assign any housekeeping CPUs, especially for single NUMA node setups. We actually default to no housekeeping CPUS unless we are over a certain core count or numa node count (Although this is specific to the certification process)

I introduced this mainly because we're considering these containers for a OCP test suite and possibly for the cert-tool suite. My intention is to simplify things for the user. This tool is designed to be a versatile topology generator that adheres to best practices in most situations. Of course, there might be unique requirements, like in the telco sector, that diverge from these practices. I hadn't considered that case as much, so we probably need a way to override this if we where to use it.

It's all about user flexibility. Some users might not want to assign more than one core for housekeeping duties. And I don't believe we should rigidly assign it to the system's first CPU (Although I think this should be an option). For example tasks being performed on a separate NUMA node—they'd require housekeeping too. If, for instance, the second node lacks dedicated housekeeping, housekeeping tasks on it might get shifted to the first node's housekeeping CPU. And if you have a system with a lot of numa nodes, one housekeeping may not be enough to handle the load. It really boils down to where tasks are distributed across all CPUs.

If this isn't fitting, I'm open to revising or even splitting it into another PR. Or if its not fitting in this repo at all, I can take the logic and store it somewhere ele. There's also room to introduce a setting for those who prefer a single housekeeping node. But multiple housekeepers for multipe NUMA have their place. For instance, if we aim to push a system to its limits: imagine half the CPUs on each NUMA node creating workloads while the other half measures performance. But this may not be needed for your usecase.